### PR TITLE
Fix for update read_count in email entity.

### LINKED ISSUE #13806: Read email statistics in the Emails section within the Channels menu are never updated
### Mautic Version

5.0.x series

### Way of installing

Other (please specify if relevant in the bug report)

### PHP version

8.0.20 & 8.0.30

### What browsers are you seeing the problem on?

Not relevant

### What happened?

When we send, for example, a campaign to different contacts with a pixel in the email and then check the reading statistics in the Emails section within Channels in the dashboard, we verify, after confirming that the email was opened and loaded , that emails read are always zero. We checked in the database that all the data was correct within the email_stats table but the read_count within the emails table was not updating, we also checked everything we read about this type of issue on your forum including possible block or not loads of the pixel by a system (We think it is meaningless if email_stats is populated correctly). After checking everything, we analyzed the code and found that the new incrementRead function within the EmailRepository in the Mautic 5.x series had what we believe to be a bug, because it appears to be the cause of the problem.

We can confirm on line [628](https://github.com/mautic/mautic/blob/ac21ec48f468e5c374224971b2e2ad92d13f695f/app/bundles/EmailBundle/Entity/EmailRepository.php#L628) within EmailRepository: $q->expr()->notIn('e.id', $subQuery->getSQL()) that the subquery always obtains the emailId that has been requested to update, but in the conditions of the the expression in the main query it is found a "notIn" that it makes the reading counter in the email table is never allowed to be updated. Then you can change that line to $q->expr()->in('e.id', $subQuery->getSQL()) to solve it. Because we understand that what you are trying to do in that expression is to verify that both the email exists as read in the email_stats table for the statistic received and the email exists in the emails table.

### How can we reproduce this issue?

Step 1: Create contacts with an email that you can check, create a segment that contains those contacts, create an email and create a campaign to send an email as action to that segment
Step 2: Execute the cron jobs of segments:update, campaings:update and campigns:trigger to sincronize all and sen the campign emails
Step 3: Open and read an email of any of the contacts
Step 4: Access into mautic dashboard to Channels->Emails and check if it has mark in the emails read something > 0 or =the number of emails opened from differents contacts

### Relevant log output

```shell
Log doesn't track any relevant data or error related with this bug.
```

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

<br /><hr>
Care about this issue? Want to get it resolved sooner? If you are a <a href='https://www.mautic.org/become-a-member-of-mautic'>member of Mautic</a>, you can add some funds to the <a href='https://opencollective.com/mautic/projects/bounties'>Bounties Project</a> so that the person who completes this task can claim those funds once it is merged by a member of the core team! Read the docs <a href='https://contribute.mautic.org/product-team/mautic-bounty-programme'>here.</a>

### LINKED ISSUE #13866: read_count in email entity never update
### Mautic Version

5.0.x series

### Way of installing

I installed with composer using https://github.com/mautic/recommended-project

### PHP version

8.1

### What browsers are you seeing the problem on?

Not relevant

### What happened?

When a mail is send out, the is_read entry in the email entity always stays zero. This number is shown on the mail overview, and in the email details tab, where it stays zero. 
Graphs do show the correct number, since they get the data from email_stats table, which is updated correctly.

The issue is AFAICT introduced in M5.1 from following PR. https://github.com/mautic/mautic/pull/12673

Going through the code path, it starts from hitEmail function in EmailModel. https://github.com/mautic/mautic/blob/5.x/app/bundles/EmailBundle/Model/EmailModel.php

Relevant part line 439-451. 
```php
$this->emailStatModel->saveEntity($stat);

        // Only up counts if associated with both an email and lead
        if ($firstTime && $email && $lead) {
            try {
                $this->getRepository()->incrementRead($email->getId(), $stat->getId(), $email->isVariant());
            } catch (\Exception $exception) {
                error_log($exception);
            }
        }
        if ($email) {
            $this->em->persist($email);
        }
```

This updates the stat, followed by updating the email entity itself (the incrementRead function).
However the increment read function uses following query. 

```php
$subQuery = $this->getEntityManager()->getConnection()->createQueryBuilder()
            ->select('es.email_id')
            ->from(MAUTIC_TABLE_PREFIX.'email_stats', 'es')
            ->where('es.id = :statId')
            ->andWhere('es.is_read = 1');

        $q->update(MAUTIC_TABLE_PREFIX.'emails', 'e')
            ->set('read_count', 'read_count + 1')
            ->where(
                $q->expr()->and(
                    $q->expr()->eq('e.id', ':emailId'),
                    $q->expr()->notIn('e.id', $subQuery->getSQL())
                )
            )
            ->setParameter('emailId', $emailId)
            ->setParameter('statId', $statId);

        if ($isVariant) {
            $q->set('variant_read_count', 'variant_read_count + 1');
        }
```

Where it only updates the read_count for mails not yet read in email_stats. Which was done just before we get into here, so we never actually update it. (This filter because of not counting mails being read multiple times). 

Might be enough to switch the stat saving to after the email entity update, but not sure about that. 

### How can we reproduce this issue?

Step 1: Create a mail. 
Step2: Send out mail.
Step 3: Read the mail and see the email is_read column not updated. 

### Relevant log output

_No response_

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

<br /><hr>
Care about this issue? Want to get it resolved sooner? If you are a <a href='https://www.mautic.org/become-a-member-of-mautic'>member of Mautic</a>, you can add some funds to the <a href='https://opencollective.com/mautic/projects/bounties'>Bounties Project</a> so that the person who completes this task can claim those funds once it is merged by a member of the core team! Read the docs <a href='https://contribute.mautic.org/product-team/mautic-bounty-programme'>here.</a>

### PR DESCRIPTION
**Pull request by abhisekmazumdar**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🟢 
| New feature/enhancement? (use the a.x branch)      | 🔴
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🔴
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | https://github.com/mautic/mautic/issues/13866 

## Description

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a mail.
3. Send out mail.
4. Read the mail and see the email is_read column not updated.

### CODE REVIEW

**Comment by pm-pmaas** [comment]
@RCheesley i can confirm that everything is working as expected after this update:
Scherm­afbeelding 2024-07-04 om 11 01 56

We still need a solution to also fix the historic send e-mails, to set totals again, see topic: [](https://github.com/mautic/mautic/issues/13866)

**Comment by RCheesley** [comment]
@pm-pmaas can you please go to files changed > leave review (documented [here](https://contribute.mautic.org/contributing-to-mautic/tester#leaving-your-review)) so that it counts as a review, rather than just a comment? Thanks!

**Review by pm-pmaas — CHANGES_REQUESTED** [review]
I 
Scherm­afbeelding 2024-07-04 om 11 01 56
can confirm that everything is working as expected after this update:
Scherm­afbeelding 2024-07-04 om 11 01 56

We still need a solution to also fix the historic send e-mails, to set totals again, see topic:

**Comment by abhisekmazumdar** [comment]
> I Scherm­afbeelding 2024-07-04 om 11 01 56 can confirm that everything is working as expected after this update: Scherm­afbeelding 2024-07-04 om 11 01 56
> 
> We still need a solution to also fix the historic send e-mails, to set totals again, see topic:

Looks like we need to write a migration for this. I will update the PR.

**Comment by andersonjeccel** [comment]
@abhisekmazumdar This PR needs rebase too
Bug fixes should be merged into the 5.1 instead of 5.x branch

**Comment by abhisekmazumdar** [comment]
Sorry, I created a mess while trying to rebase the base branch. But now, everything looks sorted and clean.

I have updated the PR with the required migration logic. Let me know your thoughts on this.

File: app/migrations/Version20240708153845.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        $sql          = sprintf('SELECT id, read_count FROM %s', $this->emailsTableName);
+        $emailsResult = $this->connection->executeQuery($sql)->fetchAllAssociative();
+
+        foreach ($emailsResult as $email) {
+            $statsResult = $this->connection
+              ->executeQuery(
+                  "SELECT email_id, is_read 
+                        FROM {$this->emailStatsTableName} 
+                        WHERE email_id = {$email['id']}"
```
Comment (mallezie) [inline app/migrations/Version20240708153845.php]: Replacing this with a grouped count query, is probably more efficient. 
Email_stats can be huge on some mautic installs.

Comment (abhisekmazumdar) [reply app/migrations/Version20240708153845.php]: I'm still burnishing my SQL skills. Drupal 8+ didn't require me to write much SQL queries, so I'm working on improving that.

I have updated the PR. I hope this looks good to you.

**Review by pm-pmaas — APPROVED** [review]
I can confirm that this fix works for updating the history e-mails with the correct stats.

**Comment by RCheesley** [comment]
Thanks so much for your work on this @abhisekmazumdar !

File: app/migrations/Version20240708153845.php (hunk 1 above)
Comment (RCheesley) [reply app/migrations/Version20240708153845.php]: @mallezie could you re-review from the code perspective please? Thanks!

File: app/migrations/Version20240708153845.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            $totalCountResult = $this->connection
+              ->executeQuery(
+                  "SELECT email_id, SUM(is_read) as total_read_count 
+                        FROM {$this->emailStatsTableName} 
+                        WHERE email_id = :email_id 
+                        GROUP BY email_id",
+                  ['email_id' => $email['id']]
+              )
+              ->fetchAssociative();
+            if ($email['id'] === $totalCountResult['email_id'] && (int) $email['read_count'] < $totalCountResult['total_read_count']) {
```
Comment (AlanWierzchonCA) [inline app/migrations/Version20240708153845.php]: Can we add query result checking here, e.g.::
```
if (is_array($totalCountResult) && $email['id'] === $totalCountResult['email_id'] && (int) $email['read_count'] < $totalCountResult['total_read_count']) {
...
```

?

**Review by AlanWierzchonCA — CHANGES_REQUESTED** [review]

**Comment by ralph9001** [comment]
Hi,
regarding to this https://github.com/mautic/mautic/pull/13915/files I changed app/bundles/EmailBundle/Model/EmailModel.php and create app/migrations/Version20240708153845.php

But my stats still not update on on new emails, what else i need to do?

File: app/migrations/Version20240708153845.php (hunk 2 above)
Comment (abhisekmazumdar) [reply app/migrations/Version20240708153845.php]: Yes I see that will be helpful.

**Comment by abhisekmazumdar** [comment]
> Hi, regarding to this https://github.com/mautic/mautic/pull/13915/files I changed app/bundles/EmailBundle/Model/EmailModel.php and create app/migrations/Version20240708153845.php
> 
> But my stats still not update on on new emails, what else i need to do?

Did you run the migration?

**Review by AlanWierzchonCA — APPROVED** [review]

**Review by kuzmany — APPROVED** [review]
Seems good 👍

**Pull request merged**